### ISSUE TITLE
Add support for java agents in test tasks/suites

### ISSUE THREAD

**Issue by Sineaggi** [gradle/gradle#27714]
### Expected Behavior

Tests run without warnings from the jvm.

Perhaps the expectation would be to add test agent dependencies to a new testAgent configuration (that would map to `-javaagent:/path/to/agent.jar`). Maybe in the new test-suites plugin?

For example, something like

```kotlin
testing {
    suites {
        val test by getting(JvmTestSuite::class) {
            useJUnitJupiter("5.10.1")
            dependencies {
                implementation("io.mockk:mockk:1.13.9")
                testAgent("net.bytebuddy:byte-buddy-agent:1.14.6")
            }
...
```

### Current Behavior (optional)

In java 21, agents warn when loaded dynamically.
```
WARNING: A Java agent has been loaded dynamically (/Users/cwalker/.gradle/caches/modules-2/files-2.1/net.bytebuddy/byte-buddy-agent/1.14.6/46e2545d7a97b6ccb195621650c5957279eb4812/byte-buddy-agent-1.14.6.jar)
WARNING: If a serviceability tool is in use, please run with -XX:+EnableDynamicAgentLoading to hide this warning
WARNING: If a serviceability tool is not in use, please run with -Djdk.instrument.traceUsage for more information
WARNING: Dynamic loading of agents will be disallowed by default in a future release
```

### Context

Using mockk, which loads the bytebuddy agent dynamically.

**Comment by cobexer** [comment]
Thank you for your interest in Gradle!

This feature request is in the backlog of the relevant team and is prioritized by them.
***
This feature is about having a higher level API for java agents in tests, it is not about Gradle hiding the warning.

**Comment by xtermi2** [comment]
How is the progress here?
We have a similar workaround, but this destroys the cacheability when using `jvmArgs` with an absolute path!
Or is there an "cache friendly" workaround to set an javaagent in the `Test` task without?

**Comment by xtermi2** [comment]
> Or is there an "cache friendly" workaround to set an javaagent in the `Test` task?

I've found one:
```kotlin
val testAgent = configurations.maybeCreate("testAgent")
testAgent.isTransitive = false
dependencies.add(testAgent.name, "org.aspectj:aspectjweaver:1.9.22.1")
val downloadJavaAgentTask = tasks.register(
                 "downloadJavaAgent",
                Copy::class.java
            ) {
                from(testAgent)
                into(layout.buildDirectory.dir("testAgent"))
                rename { "aspectjweaver.jar" }
                duplicatesStrategy = DuplicatesStrategy.FAIL
            }

// in the Test task
dependsOn(downloadJavaAgentTask)
val aspectjweaverRelativePath = layout.buildDirectory.file("testAgent/aspectjweaver.jar").get().asFile.toRelativeString(projectDir)
jvmArgs("-javaagent:$aspectjweaverRelativePath")
```

**Comment by TimvdLippe** [comment]
We have a long thread at Mockito on our support for Java 21: https://github.com/mockito/mockito/issues/3037 We shipped a java agent and it now works. While we solved the problem on our end, we lately have been hearing from a lot of our users that the integration with build tooling is not good enough. In fact, @jprinet sent us https://github.com/mockito/mockito/pull/3606 where our documentation is not optimal for Gradle build performance.

There have been numerous GitHub discussions (https://github.com/mockito/mockito/discussions/3603 https://github.com/mockito/mockito/discussions/3596) as well as community maintained plugins (https://github.com/ryandens/javaagent-gradle-plugin?tab=readme-ov-file#java-test-task-integration https://plugins.gradle.org/plugin/com.bakdata.mockito) that all attempt to tackle this problem.

At this point, given that solutions diverge, have various issues/inconsistencies/non-performance, I would propose this to be tackled by Gradle itself. This is especially important given that the IDE doesn't always support any of these plugins (https://github.com/gradle/gradle/issues/16841 https://github.com/mockito/mockito/issues/3037#issuecomment-2569680756 https://stackoverflow.com/questions/77951485/getting-a-java-agent-has-been-loaded-dynamically-warning-in-intellij-after-upg/79213258#79213258).

Please consider this data as an opportunity to streamline java-agent usage (whether it is Mockito or a different project) in Gradle.

**Comment by tkrah** [comment]
JFTR, until this is maybe resolved in gradle itself, there is a gradle plugin you can use in the meantime which does add support for the agent:

https://github.com/ryandens/javaagent-gradle-plugin?tab=readme-ov-file#java-test-task-integration

as another workaround.

**Comment by bric3** [comment]
I agree with @TimvdLippe, I'll even add that supporting passing dependencies as `-javaagent:...` could be useful for many Gradle "configurations", not just tests, so custom configurations like integration tests, etc, could also benefit from it.

**Comment by ryandens** [comment]
👋 Hey folks - happy to take recommendations from the community on https://github.com/ryandens/javaagent-gradle-plugin/issues/225 or to make some contributions to gradle in this area (if desired). The nature of Java agents and their configurations can make it hard to share dependencies via a platform. I'm thinking using a component metadata rule may be the best approach here, but I'm still figuring out what that looks like in practice!

**Comment by TimvdLippe** [comment]
For those folks coming here from Mockito and wondering why we are not linking to a specific Gradle plugin, please read our rationale here: https://github.com/mockito/mockito/issues/3705#issuecomment-3174480393

> Currently, we don't want to "bless" a specific Gradle plugin, since they all work slightly differently. Until there is convergence on a solution and one "blessed" by the Gradle maintainers themselves, we prefer not to link to a single solution.

**Comment by Sineaggi** [comment]
I don't know if it makes more sense to add agent support to test suites directly, or if it makes sense to add agent support everywhere, `runtimeOnlyAgent`, `testRuntimeOnlyAgent` that way devs who are using tools like the [datadog](https://central.sonatype.com/artifact/com.datadoghq/dd-java-agent), [contrast](https://central.sonatype.com/artifact/com.contrastsecurity/contrast-agent) or [newrelic](https://central.sonatype.com/artifact/com.newrelic.agent.java/newrelic-agent) agent jars. Perhaps there's a way to generify this for all agent users.
